subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClasspathUtil.java

```diff
@@ -63,7 +63,7 @@ public void visitClassPath(URL[] classPath) {
                         try {
                             implementationClassPath.add(new File(toURI(url)));
                         } catch (URISyntaxException e) {
-                            throw new UncheckedException(e);
+                            throw UncheckedException.throwAsUncheckedException(e);
```

Comment:
This is actually the reverse of what Google is suggesting in https://github.com/google/guava/wiki/Why-we-deprecated-Throwables.propagate. Why do it like this?

My main issue is that this now mashes together code paths designed to handle arbitrary exceptions including thread interrupts with handling an exception that has nothing to do with that.

Reply:
I've simply removed the public constructor to make sure that all code paths use the safe behavior. Killing UncheckedException would be a whole ordeal in itself, beyond the scope of this PR:

Reply:
I don't see how this change makes the code any better now, or how it gets us closer to not require an `UncheckedException`. Getting rid of `UncheckedException` is not my goal, as there are cases where you need something like that.

If your goal is to avoid having `InterruptedException` be re-thrown without the thread being interrupted again, you could have a separate constructor that takes an `InterrputedException` and deprecate it, redirecting use to `Interrputs.rethrowAsUncheckedAndInterruptCurrentThread()` or something. Even interrupting the custom constructor would be better.

Reply:
I think a realistic goal could be to eliminate the use of unchecked exceptions for interruptions (and let them bubble up instead). Combining code paths involving `InterruptedException` with other exceptions actually makes it harder to track down these uses, not easier.

Reply:
You are assuming we always statically know we're dealing with InterruptedException.

There are almost 200 places where we catch Throwable and I have no idea which of those are also catching InterruptedException without touching all of those as well. These are now all handled by changing the behavior of this rethrow method. Getting rid of that is just too much for this PR.

I wanted to have one consistent behavior before tackling the overarching problem of exception handling in general.